src/components/matrix/row-tree-headers.tsx

```diff
@@ -0,0 +1,548 @@
+    listData, 
+    setFilteredGridYData, 
+    setVisiableRowNodes,
+    isScrolling } = props.itemData;
+
+  const [prevSearched, setPrevSearched] = useState<string | null>(null); // previous searched enrty
+  const [expanded, setExpanded] = useState<string[]>([]); // all expanded nodes
+  const [treeDataDict, setTreeDataDict] = useState<Record<string, MatrixTreeDatum>>({}); // Dictionary to store relationship of parent and child for tree data
+
+  /**
+   * styles for tree view and tree items
+   */
+  // the font size is 14 for this page
+  const iconSize = 14;
+
+  // style for the whole tree
+  const rowTreeHeadersStyles: CSSProperties = {
+    position: 'absolute',
+    direction: 'rtl',
+    top: props.top,
+    height: props.height,
+    width: props.width,
+    overflow: 'auto',
+    willChange: 'transform',
+  };
+
+  /**
+   * Interaction functions
+   */
+
+  // Check whether all ancestors of a node exist in visitedNode list
+  const checkParentChainExist = (treeDataDict: Record<string, MatrixTreeDatum>, nodeId: string, visitedNodes: Set<string>): boolean => {
+    const node = treeDataDict[nodeId];
+    if (node.parent_id === null) {
+      return true; // Reached the top of the chain
+    }
+    if (!visitedNodes.has(node.parent_id)) {
+      return false; // Detected the parent node does not exist in visitedNodes
+    }
+    return checkParentChainExist(treeDataDict, node.parent_id, visitedNodes);
+  };  
+
+  // Find all ancestor nodes of the searched node
+  const getParentChain = (treeDataDict: Record<string, MatrixTreeDatum>, nodeId: string, visitedNodes: Set<string>): Set<string> => {
+    let node = treeDataDict[nodeId];
+    if (node.parent_id === null) {
+      return visitedNodes; // Reached the top of the chain
+    }
+    node = treeDataDict[node.parent_id];
+    while (node && node.child_id !== null) {
+      visitedNodes.add(node.child_id);
+      node = treeDataDict[node.parent_id];
+    }
+    return visitedNodes;
+  };
+
+  // Initialize the dictionary to store relationship of parent and child for tree data
+  useEffect(() => {
+    const initialTreeDataDict = props.treeData.reduce((dict: Record<string, MatrixTreeDatum>, node: MatrixTreeDatum) => {
+      dict[node.child_id] = node;
+      return dict;
+    }, {});
+    setTreeDataDict(initialTreeDataDict);
+  }, [props.treeData]);
+
+  /**
+   * Refresh visiableRowNodes and filteredGridYData whenever expanded nodes change,
+   * update previous Searched entry as well.
+   * This means the header and grid data should sync whenever expanded nodes change
+   */
+  useEffect(() => {
+    const newSet = new Set<string>();
+    // Add all top nodes by default
+    for (const node of props.treeNodes) {
+      newSet.add(node.key);
+    }
+    // Check visibility and add all visiable nodes to the set
+    const nodesSet: Set<string> = new Set(expanded);
+    expanded.forEach(nodeId => {
+      const visiable = checkParentChainExist(treeDataDict, nodeId, nodesSet);
```

Comment:
Please fix this typo. Also let's use the similar `if () return` that I mentioned above to make this easier to read.

```tsx
const isVisible = checkParentChainExist(treeDataDict, nodeId, nodesSet);
if (!visible) return;

const node = props.treeNodesMap[nodeId];
if (!node || node.children.length === 0) return;

...
```